src/OpenTelemetry.Api.ProviderBuilderExtensions/CHANGELOG.md

```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Fixed the `TracerProviderBuilder.AddInstrumentation` factory extension.
```

Comment:
Can you describe more or give an example of what happened if you used this before the fix? I'm assuming `builder` is a different instance than `tracerProviderBuilder`, the calling `AddInstrumentation` had no effect?

Reply:
Can't tell if you are asking me to explain that here or in the CHANGELOG 🤣 I'll start with just here...

> I'm assuming builder is a different instance than tracerProviderBuilder, and so calling AddInstrumentation had no effect?

Correct! Calling it before this fix it will just no-op.

Reply:
So `tracerProviderBuilder` instance is not `ITracerProviderBuilder`, but the `builder` instance is? Then, what about the [same check](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4468/files#diff-273272aa8fc7804bd10d63c71ab39b3bd0fd968e11a0e2ab708daf19027af69dR134) in `ConfigureServices`? Do we need to worry about that?

Reply:
There's 2 builders...

* `TracerProviderBuilderBase` this is the phase 1 builder. Implements `ITracerProviderBuilder`. Dumps everything into the `IServiceCollection`.

* `TracerProviderBuilderSdk` this is the phase 2 builder which holds the state. Everything from phase 1 is re-played against this builder once `IServiceProvider` is available in the order it was registered. The state then gets consumed into the `TracerProviderSdk`.

The working version of the clause is...

```csharp
            if (builder is ITracerProviderBuilder iTracerProviderBuilder
                && iTracerProviderBuilder.Provider != null)
```

* The first part (`is ITracerProviderBuilder iTracerProviderBuilder`) is true for both `TracerProviderBuilderBase` & `TracerProviderBuilderSdk`.

* The second part (`iTracerProviderBuilder.Provider != null`) is the more interesting one! Only `TracerProviderBuilderSdk` has access to the `IServiceProvider`. During phase 1 we operate on the `IServiceCollection`. During phase 2 `IServiceCollection` has been closed and we have the `IServiceProvider`.

The way the code was written before it was capturing the "phase 1" builder (`TracerProviderBuilderBase`) so the second clause would evaluate to `false` even when the final "phase 2" builder was in play.

Yes, this is confusing 😢

Reply:
> Can't tell if you are asking me to explain that here or in the CHANGELOG 🤣 I'll start with just here...

Both 😆, appreciate the explanation here, but I also think elaborating in the changelog would be good as it might help users determine if they're impacted.

It does not appear that any of our instrumentation is using this particular overload of `AddInstrumentation`, so I'd imagine that end users who may be impacted by this bug are ones that are using this overload in some kind of DI scenario.

Reply:
Updated the CHANGELOG text to be more specific. LMK if you want it to be more verbose. I'll also spin up an issue for this and link to it from the PR description.